detection_rules/ml.py

```diff
@@ -0,0 +1,445 @@
+from .misc import client_error
+from .schemas import definitions
+from .utils import get_path
+
+
+ML_PATH = Path(get_path('machine-learning'))
+
+
+def info_from_tag(tag: str) -> (Literal['ml'], definitions.MachineLearningType, str, int):
+    try:
+        ml, release_type, release_date, release_number = tag.split('-')
+    except ValueError as exc:
+        raise ValueError(f'{tag} is not of valid release format: ml-type-date-number. {exc}')
+
+    return ml, release_type, release_date, int(release_number)
+
+
+class InvalidLicenseError(Exception):
+    """Invalid stack license for ML features requiring platinum or enterprise."""
+
+
+@dataclass
+class MachineLearningClient:
+    """Class for experimental machine learning release clients."""
+
+    es_client: Elasticsearch
+    bundle: dict
+
+    @cached_property
+    def model_id(self) -> str:
+        for name, data in self.bundle.items():
+            if Path(name).stem.lower().endswith('model'):
+                return data['model_id']
+
+    @cached_property
+    def bundle_type(self) -> str:
+        return self.model_id.split('_')[0].lower()
+
+    @cached_property
+    def ml_client(self) -> MlClient:
+        return MlClient(self.es_client)
+
+    @cached_property
+    def ingest_client(self) -> IngestClient:
+        return IngestClient(self.es_client)
+
+    @cached_property
+    def license(self) -> str:
+        license_client = LicenseClient(self.es_client)
+        return license_client.get()['license']['type'].lower()
+
+    def verify_license(self):
+        valid_license = self.license in ('platinum', 'enterprise')
+
+        if not valid_license:
+            err_msg = 'You must have a platinum or enterprise subscription in order to use these ML features'
+            raise InvalidLicenseError(err_msg)
+
+    @classmethod
+    def from_release(cls, es_client: Elasticsearch, release_tag: str, repo: str = 'elastic/detection-rules'
+                     ) -> 'MachineLearningClient':
+        """Load from a GitHub release."""
+        full_type = '-'.join(info_from_tag(release_tag)[:2])
+        release_url = f'https://api.github.com/repos/{repo}/releases/tags/{release_tag}'
+        release = requests.get(release_url)
+        release.raise_for_status()
+
+        # check that the release only has a single zip file
+        assets = [a for a in release.json()['assets'] if
+                  a['name'].startswith(full_type) and a['name'].endswith('.zip')]
+        assert len(assets) == 1, f'Malformed release: expected 1 {full_type} zip file, found: {len(assets)}!'
+
+        zipped_url = assets[0]['browser_download_url']
+        zipped_raw = requests.get(zipped_url)
+        zipped_bundle = zipfile.ZipFile(io.BytesIO(zipped_raw.content))
+
+        bundle = {}
+        for filename in zipped_bundle.namelist():
+            if filename.endswith('/'):
+                continue
+
+            fp = Path(filename)
+            contents = zipped_bundle.read(filename)
```

Comment:
do we already have a utility function to convert a zip into a `dict[str, bytes]`? if not, i think this could be a good idea